community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/execution/PipeExecutionPlanBuilder.scala

```diff
@@ -23,271 +23,381 @@ import org.neo4j.cypher.internal.compiler.v3_0.ast.convert.commands.ExpressionCo
+          result
+        }
+      }
 
-        case SortedLimit(lhs, exp, sortItems) =>
-          TopNPipe(buildPipe(lhs), sortItems.map(_.asCommandSortItem).toList, toCommandExpression(exp))()
+      val nodeInScope = expansionMode match {
+        case ExpandAll => false
+        case ExpandInto => true
+      }
+      VarLengthExpandPipe(source, fromName, relName, toName, dir, projectedDir,
+        LazyTypes(types), min, max, nodeInScope, predicate)()
 
-        // TODO: Maybe we shouldn't encode distinct as an empty aggregation.
-        case Aggregation(Projection(source, expressions), groupingExpressions, aggregatingExpressions)
-          if aggregatingExpressions.isEmpty && expressions == groupingExpressions =>
-          DistinctPipe(buildPipe(source), groupingExpressions.mapValues(toCommandExpression))()
+    case Optional(inner) =>
+      OptionalPipe(inner.availableSymbols.map(_.name), source)()
 
-        case Aggregation(source, groupingExpressions, aggregatingExpressions) if aggregatingExpressions.isEmpty =>
-          DistinctPipe(buildPipe(source), groupingExpressions.mapValues(toCommandExpression))()
+    case Sort(_, sortItems) =>
+      SortPipe(source, sortItems)()
 
-        case Aggregation(source, groupingExpressions, aggregatingExpressions) =>
-          EagerAggregationPipe(
-            buildPipe(source),
-            groupingExpressions.keySet,
-            Eagerly.immutableMapValues[String, ast.Expression, AggregationExpression](aggregatingExpressions, buildExpression(_).asInstanceOf[AggregationExpression])
-          )()
+    case plans.Skip(_, count) =>
+      SkipPipe(source, buildExpression(count))()
 
-        case FindShortestPaths(source, shortestPathPattern, predicates) =>
-          val legacyShortestPath = shortestPathPattern.expr.asLegacyPatterns(shortestPathPattern.name.map(_.name)).head
-          new ShortestPathPipe(buildPipe(source), legacyShortestPath, predicates.map(toCommandPredicate))()
+    case plans.Limit(_, count) =>
+      LimitPipe(source, buildExpression(count))()
 
-        case Union(lhs, rhs) =>
-          NewUnionPipe(buildPipe(lhs), buildPipe(rhs))()
+    case SortedLimit(_, SignedDecimalIntegerLiteral("1"), sortItems) =>
+      Top1Pipe(source, sortItems.map(_.asCommandSortItem).toList)()
 
-        case UnwindCollection(lhs, identifier, collection) =>
-          UnwindPipe(buildPipe(lhs), toCommandExpression(collection), identifier.name)()
+    case SortedLimit(_, exp, sortItems) =>
+      TopNPipe(source, sortItems.map(_.asCommandSortItem).toList, toCommandExpression(exp))()
 
-        case LegacyIndexSeek(id, hint: NodeStartItem, _) =>
-          val source = new SingleRowPipe()
-          val ep = entityProducerFactory.nodeStartItems((planContext, StatementConverters.StartItemConverter(hint).asCommandStartItem))
-          NodeStartPipe(source, id.name, ep, Effects(ReadsAllNodes))()
+    // TODO: Maybe we shouldn't encode distinct as an empty aggregation.
```

Comment:
There are a couple of TODOs left in this file that probably should not be there when we merge this in